DM-24172: Implement DBB Buffer Mngrs v1.1

mxk62 opened the pull request.

mxk62 committed:
Get rid of old code base

Previous version of the endpoint buffer manager was in fact `ctrl_oods`.
The new one will be a total overhaul so I removed the existing code base
leaving just files required by LSST build and package management
systems.

mxk62 committed:
Add initial implementation of Finder

mxk62 committed:
Add Delete action

mxk62 committed:
Add a function for calculating files' checksums

mxk62 committed:
Make Finder check for duplicates

Finder was moving files from the buffer to the storage area and updating
the database records unconditionally.  This led to a situation where
duplicates were possible both in the database and the storage.  To
address this, now, for each file found in the buffer, Finder:

* Calculates file's checksum.
* Query the database to find out if there is a file in the storage with
  the same checksum.
* If such a file is found, Finder removes the file from the buffer and
  proceeds to a next discovered file.
* Otherwise, it moves the file to the storage area and updates the
  database records.

mxk62 committed:
Record time at which file was added

Added time when a file was discovered to the information stored in the
database.

mxk62 committed:
Make docstring for scan strategy better

mxk62 committed:
Improve log messages in Finder

mxk62 committed:
Make a minor name adjustments in Finder

mxk62 committed:
Simplify dealing with duplicates

The part of the Finder responsible for dealing with duplicates had a
non-intuitive control flow which found a reflection in voluminous
comments.  I made it simpler which not only made the code easier to
follow, but allow me to get rid of these comments.

mxk62 committed:
Make file discovery method configurable

During the initial development there was only one usable method allowing
Finder to discover new files in a specified location and it was
hard-coded.  Once a second method became available, that needed to be
changed.  I modified the Finder to make selecting search method Finder's
configuration option.

mxk62 committed:
Add new method for file discovery

Previously, only one method, using os.walk(), existed which allow the
Finder to discover files in a specified location.  This method could
only be used in DBB buffers where each existing file is considered to be
new.  Discovering files in the storage area requires different approach.
I added a method which identifies new files  by parsing rsync log files.

mxk62 committed:
Add __repr__ to File ORM

File's ORM didn't have any printable representation.  I added __repr__
method to address this shortcoming.

mxk62 committed:
Add the framework responsible for file ingestion

By design, the endpoint buffer manager consist of two frameworks
responsible for file discovery and file ingestion to different database
systems.  Finder is an implementation of the former.  Now I added the
Ingester which is an implementation of the latter.

mxk62 committed:
Add Gen2 Butler ingestion plugin

The Ingester needs a plugin for each database system it is supposed to
ingest files to.  I added an initial implementation of the plugin which
allow Ingester to ingest files to an existing Gen2 Bulter repository.

mxk62 committed:
Add a manager driver

Added a driver responsible for configuring and starting necessary
components of the DBB endpoing buffer manager along with the example
configuration file and its schema.

mxk62 committed:
Make capitalization consistent in error messages

Capitalization in error messages was inconsistent.  Fixed it.

mxk62 committed:
Address various issues with configuration schema

There was few errors and omissions in the configuration schema.  Fixed
them and now provided example configuration can be properly validated.

mxk62 committed:
Fix various little bugs

mxk62 committed:
Fix comments in Finder run() method

mxk62 committed:
Fix problem with reporting ingest task version

The version of the LSST task used by the Gen2 Butler ingest plugin was
not being reported correctly.  Fixed it.

mxk62 committed:
Adjust time data for ingest attempts

A time and duration is stored in the database for each ingest attempts.
However, initial representation of this data turned out to be rather
poorly chosen, attempt time was represented as Unix time, duration as
seconds.  I changed attempt time to ISO date format and expressed
duration of an attempt in milliseconds.

mxk62 committed:
Save entire traceback for an ingest failure

Only the last exception's message was stored in the database when an
ingest attempt failed.  However, it turns out that saving just the last
message will, at least or Gen2 Butler, provide little to no insight what
was the reason behind the failure (currently message looks like: "Failed
to ingest file <filename>").  That I decided to save entire traceback as
it contains more helpful messages from all exceptions raised during a
failed attempt.

mxk62 committed:
Rename column in table for ingest attempts data

Column for saving version of the task which is used to perform the
ingest to the was called "version".  Made it more specific by renaming
it to "task_version"

mxk62 committed:
Fix object mappers

Existing definitions of object mappers were triggering various
SQLAlchemy errors.  Fixed them.

mxk62 committed:
Do not allow for setting incorrect file status

Ingester allow a user to select a files which it will process based on
their status.  However, this status cannot be set to value which
represent status of files which have been ingested successfully
(currently: SUCCESS) as there will be no way for the Ingester to tell
the difference between "old" and "new" successful attempts.  Thus I
added a safeguard which will raise an exception when it happens.

mxk62 committed:
Add missing docstrings and few extra comments

mxk62 committed:
Make minor changes in log messages

mxk62 committed:
Alter Ingester control flow

Ingester was doing some unnecessary work even if there were no files to
process.  I altered the control flow to address it.

mxk62 committed:
Fix incorrect query

The database query use for the Ingester to find out about new files were
not corretly implemented.  Fixed it.

mxk62 committed:
Redesign the main driver

During initial testing it turns out the driver suffers from several
shortcomings.  I redesigned to address them.  Most notable changes are:

* Added a check making sure that a user specified correct database.
* Fixed module paths.
* Used processes instead of threads to run Finders and Ingesters to
  allow for their graceful termination.
* Partitioned "monolithic" main function to smaller chunks.
* Added missing docstrings.

mxk62 committed:
Make configuration file generic

Made settings in the configuration file more generic and commented out
optional settings.

mxk62 committed:
Rename driver file

mxk62 committed:
Add missing shebang to the driver file

mxk62 committed:
Fix module paths, again

mxk62 committed:
Make 'echo' setting optional

mxk62 committed:
Add support for Gen3 Butler ingest

Added a plugin allowing an Ingester to ingest images to a Gen3 Butler
repository.

mxk62 committed:
Make Ingester handle wider range of exception

A worker function used internally by the Ingseter to handle excetptions
happaning during image ingestion was only catching RuntimeErrors.  Thus
any other exception propagated freely to the main thread causing
Ingester process to terminate.  I modified it to handle all possible
exception as it is hard to predict what exception will be thrown by LSST
tasks respnsible for the ingest process.

mxk62 committed:
Create ingest declaratives dynamically

The incoming images might need to be ingested to different database
system.  Current designed assumes each of these systems will have an
Ingester associated with it which uses its own set of database tables to
store information about ingestion process.  To address that issue I
modified the code to make it dynamically generate ORMs (declaratives)
each Ingester is using.

mxk62 committed:
Change ingest status names

Slightly alter names of the different ingest phases which will be put
in the status table by an Ingester.

mxk62 committed:
Allow for starting microservices individually

Previously a single driver was starting/stopping all microservices
required by the DBB ingest service.  Now they can be managed
individually using a uniform CLI:

    endmgr <microservice> start [OPTIONS] FILENAME

mxk62 committed:
Improve logging

Made minor tweeks to existing log messages and added few new ones.

mxk62 committed:
Make object relational mapper explicit

Functions responsible for creating object relational mappers had built
in quite a few assumptions about the names of the database tables
required by any Finders and Ingesters.  However, no matter how
reasonable these assumption may be, there cannot be taken for granted or
enforced as the code does not create database tables itself by design.
Thus, I made changes to let these tables be explicitly specified in the
configuration file.

mxk62 committed:
Pre-process exception in ingest plugins

Any exceptions occurring during an execution of a LSST task where passed
directly from the plugin to the Ingester.  Thus adding any
"pre-processing" of these error (for example, for logging purposes)
would implicitly introduced an undesired coupling between the Ingester
and a plugin implementations.  To prevent this from happening, any error
handling is now done on the plugin.  Then a plugin raises a RuntimeError
which is intercepted by the Ingester and provided error is simply logged
and/or inserted to the database.

mxk62 committed:
Detect missing file early

Occasionally, database tables can get out of sync with the actual
content of the storage area.  For, example some files can be
accidentally deleted after they were registered in the database.
Instead of relaying on external mechanisms (e.g. LSST tasks) to detect
such a situation, I made the Ingester check if file exists in the
storage area before scheduling its ingest.

mxk62 committed:
Add missing __init__.py's

Few __init__.py's were missing causing import issues.  Fixed it.

mxk62 committed:
Improve generator parsing rsync logs

Existing implementation of the generator responsible for parsing rsync
logs in search for files is transferred was to simplistic.  Firstly, it
was parsing entire log directory. Secondly, it had no mechanism allowing
to distinguish between logs that were already parsed and new ones.  I
addressed the first shortcoming by adding options which allows one to
select which subdirectory in the main log file directory needs to be
monitored.  Also, after processing a log file it now creates an empty
"guard file" indicating that a given log file was already processed
during previous iteration.

mxk62 committed:
Make search methods behave uniformly

The search methods used to scan DBB buffer had different signature than
the method parsing rsync logs.  Additionally, it was returning absolute
paths to the files in the buffer while the parse method was returning
relative paths.  To simplify using them in the higher level code, I
unified their signatures (unused options are simply ignored) and made
them both returns relative paths.

mxk62 committed:
Switch to two-table schema

After consulting with database admins I altered the database schema used
by the manager. The major change was to merge tables with file statuses
and ingests attempts into a single table which will store a complete
history of changes for each file.  There were also minor changes such as
modifying column types.

mxk62 committed:
Bring checks for required setting up to date

Changes in the schema altered the sets of required and optional
configuration settings for both the Finder and the Ingester.  I modified
all internal checks to reflect these changes.

mxk62 committed:
Make Finder catch wider range of exception

The Finder was catching general SQLAlchemy errors while interacting with
the database.  It turns out that that's not quite enough.  Sometimes
exception raised during these interaction were thrown by other modules
(e.g. psycopg2).  Hence I changed the try-except clause to catch all
possible exceptions.

mxk62 committed:
Make Finder a bit more robust

There were places in the Finder's code which could throw exception which
would not be caught.  Made a minor tweaks in the main loop control flow
and added few more guards to address this issue.

mxk62 committed:
Change the database schema

The database schema used by the manager consisted of 3 tables.  After
consulting with database I got rid of one table with redundant
information, made a minor changes to column types, and made necessary
changes it the code base to allow it to use the new schema.

mxk62 committed:
Improve documentation

Added few missing docstrings and fixed some errors in existing ones.

mxk62 committed:
Fix a trivial bug

Fixed a trivial bug causing the Finder to look into the storage
area for new files.

mxk62 committed:
Add missing license preambles

Few files were missing license preamble required by LSST.  Fixed it.

mxk62 committed:
Fix issue with creating entries for new files

Turned out SQLAlchemy query result is a tuple even if the query returns
values form a single column.  As a result, while creating database
entries for new files the field 'files_id' was being assigned the
1-element tuple instead of the expected scalar value.  This was causing
problems while using SQLlite as the database back-end, but strangely
enough not in PostgreSQL.  Fixed the issue.

mxk62 committed:
Improve docstrings

Added missing pieces in some docstrings, made parameter description more
meaningful in others.

mxk62 committed:
Rename schemas.py to validation.py

Using the word 'schema' for validating configuration file may be
confusing in a project which also uses database manageemnt systems which
also uses schemas to describe the tables in use.  To avoid such a
situation I renamed module with descriptions used to validate
configuration file.

mxk62 committed:
Make some option names more user friendly

Some of the option names were either confusing or required rather deep
knowledge how they are used in the code.  Made them hopefully
more understandable for "mere mortals" which just want to use the
manager without going deep into the weeds.

mxk62 committed:
Allow to specify connection pool

Added a setting allowing to select SQLAlchemy connection pool.

mxk62 committed:
Remove relationship between File and Event ORMs

Following SQLAlchemy docs, I defined an explicit relationship between
File and Event ORMs.  As a result the File and in turn the Finder itself
became tightly coupled with whatever table the Event represented when
the Finder was started.  It would render this relationship meaningless
for any Ingester using different file events table than the one the
Finder was associated with so I removed it.

mxk62 committed:
Improve log messages

mxk62 committed:
Make Finder/Ingester quit after validation

mxk62 committed:
Validate plugin's configuration

Different ingest plugins can (and most likely will) have different
configurations.  Because of that they were not validated as I didn't
know how to capture this differences in the JSON schema I was using to
validate configuration files.  It turned out JSON schema support
applying subschemas conditionally, so it's no longer the case.

Allow the date for search method be null

MichelleGower commented on etc/ingester.yaml:
```diff
@@ -0,0 +1,22 @@
+    config:
+      root: /data/gen2repo
+#      mode: link
+# file_status: untried
```
Not sure if comment spacing is inconsistent (comparing to logging in finder.yaml).  Is file_status under plugin, ingester or top-level?

mxk62 replied:
I will remove the comments and fix the indentation.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/finder/abcs.py:
```diff
@@ -0,0 +1,120 @@
+            ValueError(f"Cannot add {act}: not an Action.")
+        self._actions.append(act)
+
+    def pop(self, i=None):
```
Naming comment: pop functions normally return the popped item.  If the purpose really is a delete/remove function, the code would be more intuitive if named that way.

mxk62 replied:
Got it! I'll change the name of the method to `remove`.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/finder/abcs.py:
```diff
@@ -0,0 +1,120 @@
+        path : `str`
+            Path to the file.
+
+        Returns
```
Couldn't find an example quickly looking through the LSST developer guide for this.  Noticed that the docstring says returns, but there is no value returned from the method.  It modifies a class attribute.

MichelleGower replied:
And looking at other similar methods in other files, I don't see returns in the docstring for methods that don't return a value.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/finder/actions.py:
```diff
@@ -0,0 +1,235 @@
+    def __init__(self, config):
+        Action.__init__(self)
+
+        required = {"src", "dst"}
```
Multiple places in this code have config value checking.  There is a completely different config checking available via jsonschema and validate command line option.  If configs change, changes will have to be made in multiple places.  If there is nothing additional gained by these in code checks, for this version of the DBBBM, validation could always run and checks removed from the code.  If the code is checking values that are not yet covered by the jsonschema, just make a note of this for future work.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/finder/commands.py:
```diff
@@ -0,0 +1,128 @@
+    logger.info("setting up Finder...")
+    config = configuration["finder"]
+    config["session"] = session
+    config["orms"] = mapper
```
mapper isn't used anywhere else.  Why not config["orms"] = config["tablenames"]?  Also changing the key name can make debugging issues harder (e.g., can't just search code for the key used in the config file).  But if you have a particular reason for doing this relabeling, go ahead and leave it.

mxk62 replied:
When dealing with configuration, I tend to do one section at the time. So `config` is a "index" variable which holds a section I'm currently dealing with (the entire configuration is stored in `configuration`). So there's no `config["tablenames"]` as `tablenames` are defined in the previous section I already dealt with saving the settings in the `mapper` for future reference. I will change "orms" to "tablenames" to keep things consistent though.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/finder/commands.py:
```diff
@@ -0,0 +1,128 @@
+            raise RuntimeError(msg)
+        else:
+            action_config = {}
+            if name == "Move":
```
Not sure why relabelling config values for an action being used by a finder is all the way up in this commands.py file.  If not easy to just fix by putting this in the Move action class, please put on future work list.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/finder/finder.py:
```diff
@@ -0,0 +1,318 @@
+            yield path
+
+
+def parse(directory, blacklist=None, isodate=None, timespan=1):
```
Naming comment:  parse is a very generic name for this "specialized search method".

mxk62 replied:
I'll change it to `parse_rsync_logs`.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/commands.py:
```diff
@@ -0,0 +1,120 @@
+    logger.info("setting up Ingester...")
+    config = configuration["ingester"]
+    config["session"] = session
+    config["orms"] = mapper
```
Same comment as in the finder about the tablenames -> orms relabelling.

mxk62 replied:
Same as in `finder` applies.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -0,0 +1,353 @@
+        self.batch_size = config.get("batch_size", 10)
+        self.daemon = config.get("daemon", True)
+        self.pause = config.get("pause", 1)
+        self.pool_size = config.get("num_threads", 1)
```
Another place of relabelling.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -0,0 +1,353 @@
+logger = logging.getLogger(__name__)
+
+
+field_names = ['realpath', 'timestamp', 'duration', 'message', 'version']
```
Is this supposed to be relpath like other places in the code?

mxk62 replied:
No, here it's an absolute path. I'll change it to `abspath`.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/commands.py:
```diff
@@ -0,0 +1,120 @@
+        except ValueError as ex:
+            msg = f"{class_.__name__}: invalid configuration: {ex}."
+            raise RuntimeError(msg)
+        config["plugin"] = plugin
```
(Assuming I read this code right and ingester object is being instantiated here) with init methods creating objects like a Gen3 butler, there are no problems making a single object with database connections here and share it among threaded workers later?  (BTW took me a while to track this down because in the later thread code I thought config["plugin"] was still the dictionary from the config file.)

mxk62 replied:
I see your point. Indeed I was "covertly" adding, overwriting existing settings in the `finder/ingester` section of the global configuration. I'll rewrite relevant parts to make it clear what is the configuration read from the file and what is the actual configuration for the Finder/Ingester.

MichelleGower reviewed (approved):
Most comments are about minor things.  The possible big one is whether the ingesters are thread-safe.  If this only affects Gen3 ingester, then can merge these changes and add to future work list.

mxk62 committed:
Rename method removing actions from a macro

The pop() method removing an action from a macro was not retiring any
value (which a pop-like method usually does).  Changed it to remove.

mxk62 committed:
Remove obsolete section from a docstring

mxk62 committed:
Remove comments from example configurations

Comments in the configuration files were indicating which settings are
optional.  However, they made unclear which options belong to a given
(sub)section.  Hence, I removed them.

mxk62 committed:
Improve variable naming

Despite having similar role, some variables were named
inconsistently.  Also, a function name was too generic for its
specialized use.  Fixe it.

mxk62 committed:
Make dealing with configs less obscure

In the exisiting version I was "covertly" adding and redefining settings
in the configuration read from the file before passing it the
Finder/Ingester making hard distinguishing between the settings read
from the file and a configuration for a given object.  I rewrote that
making things hopefully more clear.

mxk62 committed:
Enforce interface of a Plugin

Plugins were sharing the same interface, but it wasn't enforced in any
way.  To address it, I defined a Plugin abstract base class and made
existing plugins concrete realization of it.

mxk62 committed:
Instantiate plugins late

Ingest plugins were instantiated early, i.e., before the Ingester was
even started.  However, this led to a situation where the database
connections used by the ingest code of the data management systems used
by the plugins were shared among threads.  This fact wasn't received
well by SQLite used internally by Gen2 Butler.  As it also may cause
problem with PostgreSQL, I modified to instantiate an ingest plugin in
the worker thread.  While it surely adds some overhead, this approach
should ensure that no database connection will be shared among the
threads.

Commit:
Fix setting name

Corrected the name of a setting I missed earelier to more user friendly
version.

mxk62 committed:
Update JSON schema

The name of the method for searching files by parsing rsync logs has
changed.  Updated validation schema to reflect this.

mxk62 committed:
Allow action specification to be null

When defining actions for Finder an operator needs to always specify
them, even if no action is really required.  In such a case program
expected providing a name of the no-op action, a null object called
Null.  That might cause some confusion with the null value (null).  I
modified a program to accept the regular null value instead of the name
of the actual null object.

mxk62 committed:
Fix incorrent name of the function argument

mxk62 committed:
Add missing error message to an exception

mxk62 committed:
Fix a bug related to selecting incorrect storage

Due to incorrect check, the Finder was not selecting the correct storage
area when it was set to parse rsync logs.  Fixed it.

mxk62 committed:
Fix problem with passing date to search function

It turned out PyYAML converts strings like '2020-06-29' into actual
Python datetime.date objects behind the scenes.  It wrecked havoc on
internal logic of parse_rysnc_log functions which was written with an
assumption that it receives a string representing a date and not actual
datetime.date object.  Fixed that.

mxk62 committed:
Add a user guide

mxk62 committed:
Get run number directly from an instrument

Gen3 Butler required providing a run "number" which needed to be set
manually.  Starting from LSST Stack w_2020_25 allows one to get a
default run number directly from the instrument.  Made changes to use
that feature.

mxk62 committed:
Remove insonsitency in valid statuses

Valid statues for file events used intranally were inconsitent with
values in external script for creating the database schema.  Fixed it.
